include/xtensor/xfft.hpp

```diff
@@ -0,0 +1,317 @@
+                std::size_t result = 0;
+                for (int i = 0; i < n; i++, x >>= 1)
+                    result = (result << 1) | (x & 1U);
+                return result;
+            }
+
+            template<typename E1>
+            bool is_power2(const E1& data) noexcept
+            {
+                return ((data.size() & (data.size() - 1)) == 0);
+            }
+
+            template<
+                typename T
+            >
+            auto transform_radix2(xt::xarray<std::complex<T>> data)
+            {
+                // Length variables
+                const std::size_t n = data.shape(0);
+                int levels = std::floor(std::log2(n));
+
+                // Trignometric table
+                auto i = xt::linspace<T>(0, n / 2 - 1, n / 2) * std::complex<T>(0, 1);
+                //I would like to do this but there appears to be a compiler error in CPP14
+                //Works in CPP17 and MSVC CPP14
+                //Linker error when using numeric_constants in this header
+                //auto expTable = xt::eval(xt::exp(-2 * i * ::xt::numeric_constants<T>().PI / n));
+                auto expTable = xt::eval(xt::exp(-2 * i * pi<T>() / n));
+
+                // Bit-reversed addressing permutation
+                for (std::size_t i = 0; i < n; i++)
+                {
+                    const std::size_t j = reverse_bits(i, levels);
+                    if (j > i)
+                    {
+                        std::swap(data(i), data(j));
+                    }
+                }
+
+                // Cooley-Tukey decimation-in-time radix-2 FFT
+                // Could probably be reimplemented using stride views...
+                for (std::size_t size = 2; size <= n; size *= 2)
+                {
+                    const std::size_t halfsize = size / 2;
+                    const std::size_t tablestep = n / size;
+                    for (std::size_t i = 0; i < n; i += size)
+                    {
+                        for (std::size_t j = i, k = 0; j < i + halfsize; j++, k += tablestep)
+                        {
+                            const auto temp = data(j + halfsize) * expTable(k);
+                            data(j + halfsize) = data(j) - temp;
+                            data(j) += temp;
+                        }
+                    }
+                    // Prevent overflow in 'size *= 2'
+                    if (size == n) {
+                        break;
+                    }
+                }
+                return data;
+            }
+
+            template<
+                typename T
+                >
+            auto transform_bluestein(xt::xarray<std::complex<T>> data)
+            {
+                // Find a power-of-2 convolution length m such that m >= n * 2 + 1
+                const std::size_t n = data.size();
+                std::size_t m = 1;
+                while (m / 2 <= n) 
+                {
+                    m *= 2;
+                }
+
```

Comment:
Out of curiosity: is this more efficient / better than taking the log and rounding?

Reply:
It is more efficient, but I decided to replace it with std implementations of log and pow. The current way allows the computation of the value with only integer math but is has a linear computation time proportional to n. Using pow and log2 functions it's now a fixed execution time. I suspect with long transforms using log2 and pow would become faster with the current implementation being way faster in short transforms. Ultimately this is not in the inner loop of the algorithm, so it is best to go with a more expressive solution.